### ISSUE TITLE
Ways to improve WVC

### ISSUE THREAD

**Issue by Real-Gecko** [retspen/webvirtcloud#323]
Hey there! I think we need to discuss some things about development of WVC to avoid confusion with further development. Things to discuss are:

- [ ] code formatting
- [ ] test coverage

**Comment by Real-Gecko** [comment]
I use VS Code for development and I use yapf with following settings:
```json
    "python.formatting.yapfArgs": [
        "--style",
        "{based_on_style: pep8, column_limit: 127}"
    ],
```
I also use VS Code feature "Sort imports" for sorting imports of python libs. That's why you may see a lot of code shifting in my PRs without actual change in functionality. To avoid confusion I suggest we decide which formatter we use, to keep code same for all devs.

**Comment by Real-Gecko** [comment]
Test coverage the thing I'm currently working on, slowly but surely we're getting there.

**Comment by Real-Gecko** [comment]
I think we need to perform either migration squashing or redo migrations from scratch in the end of current dev cycle, when all changes will be settled, cause we're currently getting a lot of them.
We can also try and create a way to migrate from older version of WVC to newer, maybe with management command.

**Comment by Real-Gecko** [comment]
We may create shell script for new installations and move all pre install tasks to there, like creation of `admin` user, maybe add ability to set some settings without actually editing `settings.py` etc.

**Comment by catborise** [comment]
hi @Real-Gecko i am a system engineer more than software developer. I never worked with a developer team, because of that i am not experienced with team working. but i prefer/tend to experience it. i can follow your guidance.

**Comment by catborise** [comment]
migration squashing with after some period will be fine.(i did it with python2-->3 conversion.)

i moved setttings to appsettings page. for now settings.py is slimmed down.

i move them appsettings but your admin app will be more suitable for that configs. it can be moved... (i am creating settings with setting.objects.using(db_alias).bulk_create migration but there should be more suitable way. i need help with that :) )

**Comment by catborise** [comment]
> We may create shell script for new installations and move all pre install tasks to there, like creation of `admin` user, maybe add ability to set some settings without actually editing `settings.py` etc.

after the lastest updates this approach(moving all pre install tasks) may be more suitable way.

i have never write coverage tests, i do not now how to... i should work on it.. i need time.

**Comment by Real-Gecko** [comment]
> i am a system engineer more than software developer

What IDE do you use? Does it integrate with [YAPF](https://github.com/google/yapf) if so than everything is easy. I created `conf/requirements-dev.txt`, it adds `yapf`, `coverage` and `django-debug-toolbar` to existing requirements.

YAPF is used by me to auto format source files on save, really helps to keep code readable. However other devs may prefer other ways to format code, that's why I decided to discuss the issue. Settings for yapf I use ar in post 2. Try it out.

Coverage is used to run tests with coverage info, really helps to see what parts of your code are not tested. I use following command to run tests:
```shell
venv/bin/coverage run --source '.' --omit 'venv/*' manage.py test
```
And this one to see the results:
```shell
venv/bin/coverage html && firefox ${workspaceFolder}/htmlcov/index.html
```
You can use any browser you want instead of FF.

And finally [Debug Toolbar](https://django-debug-toolbar.readthedocs.io/en/latest/installation.html) really helps too. Installation and set up is pretty easy too.

**Comment by Real-Gecko** [comment]
> i moved setttings to appsettings page. for now settings.py is slimmed down.

Yeah, that's cool, ability to change settings on the fly is always better than messing with settings.py.


> i move them to appsettings but i think your admin app will be more suitable for that configs. it can be moved...

Well that's not hard to do: one view, one model, one template. Let's do this


> i am creating settings with setting.objects.using(db_alias).bulk_create migration but there should be more suitable way. i need help with that :)

Populating data with migrations is acceptable, but Django devs recommend [using fixtures](https://docs.djangoproject.com/en/2.2/howto/initial-data/), imo it's more flexible way, it will require some work during testing, but that's not something I would worry about.

**Comment by Real-Gecko** [comment]
> i have never write coverage tests, i do not now how to... i should work on it.. i need time.

It sounds harder [than it actually is](https://docs.djangoproject.com/en/2.2/topics/testing/overview/). If you read tests I created you'll see they're pretty simple, but help a lot. Here's computes index page test:
```python
    def test_index(self):
        response = self.client.get(reverse('computes'))
        self.assertEqual(response.status_code, 200)
```
Simply `GET` index page and check that response status is 200.

**Comment by catborise** [comment]
before that i was using pycharm, it is really nice ide for python(license is over and community edition lack of some capabilities). 
Now i started using vscode. i can add yapf settings for sync.

**Comment by Real-Gecko** [comment]
I see a lot of views doing more than one task which makes them hard to maintain and test. I think we need to split views functionality like so: one view - one task. For example list users is one view, create user is another, update is another etc.

**Comment by catborise** [comment]
i do not know why retspen choose that approach but it is useful to slim down urls otherwise.
for me, it is fine to seperate small ones but instances/views.py can be painful to split :)

**Comment by Real-Gecko** [comment]
`urls.py` is just like route map user request can take to receive required answer, so it's ok for it to be big, but big views that do more than one job is bad imo.

I'm currently working on computes where I've already split some functionality to smaller views. Instances are on the way.

**Comment by Real-Gecko** [comment]
Correct me if I'm wrong: we establish new connection to libvirt host every time there's any operation on instance. Am I right?

**Comment by Real-Gecko** [comment]
Never mind, look like I've figured it out.

**Comment by catborise** [comment]
i think yes it start new connection everytime but with the help of keep alive protocol connections are cached. libvirt service on the host keep up open the connection to serve requests fastly...

**Comment by Real-Gecko** [comment]
Yep, there's wvmConnectionManager class that pools all connections. I was just confused by wvmInstance close method.

**Comment by Real-Gecko** [comment]
I'm currently working on instances app, some serious optimizations in progress. Please avoid any work in this app, to prevent source code clashing during PR.

**Comment by catborise** [comment]
👍 
currently i am not working on webvirtcloud master, be relax :) 
be careful instances app is a behemoth.  :)

**Comment by Real-Gecko** [comment]
Yep, but it's worth it, unoptimized compute instances view: more than 5 seconds to response, 41 database requests.
![Screenshot_20200618_153108](https://user-images.githubusercontent.com/2231969/85004399-539a4600-b179-11ea-9baa-abc43e6a8390.png)

Optimized: 1.5 seconds, 2 database queries.
![Screenshot_20200618_153143](https://user-images.githubusercontent.com/2231969/85004459-657be900-b179-11ea-9aac-c5beea61c720.png)

**Comment by catborise** [comment]
wow 41 query what the heck it that. it should be max five or six queries. 1. get compute 2.get instances, 3.update records, 4.check duplicates, 5.add missings, something must be out of control.

i found it:  def refresh_instance_database(comp, inst_name, info) loops for all instances and update all of them. that causes making queries relative to instance count.
i am wondering how to solve that.

**Comment by Real-Gecko** [comment]
Currently I replaced it with this code:
```python
def refr(compute):
    domains = compute.proxy.wvm.listAllDomains()
    domain_names = [d.name() for d in domains]
    # Delete instances that're not on host
    Instance.objects.filter(compute=compute).exclude(name__in=domain_names).delete()
    # Create instances that're not in DB
    names = Instance.objects.filter(compute=compute).values_list('name', flat=True)
    for domain in domains:
        if domain.name() not in names:
            Instance(compute=compute, name=domain.name(), uuid=domain.UUIDString()).save()
```
But I don't like it too, cause it does not connect instances from WVM with domains on host.
If we try to connect 'em will get query hell again. So currently I think about workaround.

One solution is to get Instance objects from DB and work with it, but that way we don't get VM status on host.

**Comment by Real-Gecko** [comment]
OK, I think I know what to do with VM status. I added `wvmInstance` to `Instance` model that we can use to get info about domain directly via libvirt:
```python
    @cached_property
    def proxy(self):
        return wvmInstance(
            self.compute.hostname,
            self.compute.login,
            self.compute.password,
            self.compute.type,
            self.name,
        )
```
Result is looking good:
```python
>>> i = Instance.objects.get(name='test')
>>> i.proxy.instance.info()
[5, 4194304, 4194304, 2, 0]
>>> 
```

**Comment by catborise** [comment]
i think you will seperate allinstances and host only intances list refreshing.

there is not need to check duplicate uuid and names on one host. but we should handle that among other hosts.
duplicate instance definition on other hosts could cause unwanted problems. 

proxy with @cached_property is a nice approach. i like it.

**Comment by Real-Gecko** [comment]
> duplicate instance definition on other hosts could cause unwanted problems.

Actually I think no, even if there will two VMs with same UUID, but on different hosts it will not cause trouble. Cause this VMs are well on different hosts :D

But current code definitely troublesome:
```python
    instances = Instance.objects.filter(name=inst_name)
    if instances.count() > 1:
        for i in instances:
            user_instances_count = UserInstance.objects.filter(instance=i).count()
            if user_instances_count == 0:
                addlogmsg(request.user.username, i.name, _("Deleting due to multiple(Instance Name) records."))
                i.delete()
```
As it does not take into account the fact that there could be instances with the same name on different computes.

**Comment by catborise** [comment]
yes same name could cause confusion but doesn't same uuid cause migration prevention?

you are right unique name and uuid can be provided efficiently only within a cluster. these actions not a complete solution but only trivial checking

**Comment by Real-Gecko** [comment]
> yes same name could cause confusion but doesn't same uuid cause migration prevention?

Yep it's possible even though possibility of UUID clashing is really low, however I think in this case libvirt will throw some kind of error we will show to user, and then that's admin's task to deal with this tricky situation. In our case it's important to keep our DB in sync with hosts.

**Comment by Real-Gecko** [comment]
I've pushed my [WIP Instances branch](https://github.com/Real-Gecko/webvirtcloud/tree/instances) into forked repo for testing and fun. Anyone willing to test and suggest is free to do so.
WARNING: heavy WIP, not for production use.

**Comment by Real-Gecko** [comment]
I `base.html` to contain this:
```html
    {% include 'navbar.html' %}

    <div role="main" class="container">
      <div class"row">
        <div class="col-sm-12">
          <div class="float-right">
            {% block page_header_extra %}
            {% endblock page_header_extra %}
          </div>
          <h3 class="page-header">
            {% block page_header %}
            {% endblock page_header %}
          </h3>
        </div>
      </div>
      {% bootstrap_messages %}
      {% block content %}{% endblock %}
    </div>
```
So there will be no need to write page header in every template, things like search bar can go to `page_header_extra` and all messages added by `django.contrib.messages` framework will go below page header.

```html
{% block page_header %}{% trans "Instances" %}{% endblock page_header %}

{% block page_header_extra %}
{% if request.user.is_superuser %}
    {% include 'create_inst_block.html' %}
{% endif %}
{% if all_host_vms or all_user_vms %}
    <input id="filter" class="form-control" type="text" placeholder="{% trans 'Search' %}">
{% endif %}
{% endblock page_header_extra %}

{% block content %}
...
```

**Comment by catborise** [comment]
although effort or code length does not much change, if alignment problems does not occur, this way is more elegant than current.  i like it

**Comment by Real-Gecko** [comment]
> although effort or code length does not much change, if alignment problems does not occur, this way is more elegant than current. i like it

This is just an example how things can be optimized, not a final version, anyone willing to suggest improvements is free to do so.

**Comment by catborise** [comment]
as far as i see grouped view for index page is removed, practically. 
group view:
![image](https://user-images.githubusercontent.com/1725643/85253413-c7349f80-b466-11ea-8337-f3ebe5ff0d97.png)
nongrouped view:
![image](https://user-images.githubusercontent.com/1725643/85253491-f3e8b700-b466-11ea-9172-a861be526a76.png)

` all_host_vms = Instance.objects.all().prefetch_related('userinstance_set')` does not contain host info, doesn't it?  
what do you think about that?

**Comment by Real-Gecko** [comment]
I think replacing it with either [accordion](https://getbootstrap.com/docs/4.5/components/collapse/#accordion-example) or tabs which will contain tables with instances. Current implementation does not work right anyway, try sorting the table by name or RAM you'll see what I mean.

**Comment by Real-Gecko** [comment]
Accordion does not look bad IMO

![Screenshot_20200622_124223](https://user-images.githubusercontent.com/2231969/85256545-3c639d00-b486-11ea-9d80-7d7495084c2c.png)

**Comment by catborise** [comment]
i added grouped view because there is not any info about host usage. especially ram usage. This info very valuable to get idea host load. other than that already nongrouped view is like that.

**Comment by Real-Gecko** [comment]
I'll commit soon to instances branch, you'll see how it looks

**Comment by Real-Gecko** [comment]
I've pushed changes de20f810a27ab528aff6699a33ab7fbb607be4bf, not the best looking for right now, but you get the idea, most important is the optimization of templates and DB queries.

**Comment by catborise** [comment]
i think, database queries not criitical. query delay impact is very low compare to libvirt calls. focusing on queries would not affect very much.

**Comment by Real-Gecko** [comment]
Actually, as far as I can tell from testing rewritten views DB queries impact is huge :)

**Comment by catborise** [comment]
i cannot run your branch. gives error
always i get that error

![image](https://user-images.githubusercontent.com/1725643/85273228-fdcde280-b485-11ea-9740-516a007d33f0.png)

**Comment by Real-Gecko** [comment]
Is it the same error as in 2c8f092992bf657977c268e806a300cee3e30fce? If so, try commenting out `'webvirtcloud.middleware.ExceptionMiddleware',` in your `settings.py` and see where error is.

**Comment by catborise** [comment]
not reachable host caueses, i think. trying to get cpu from host info causes that. 
![image](https://user-images.githubusercontent.com/1725643/85278141-3329fe80-b48d-11ea-9e67-013f302d8d12.png)

**Comment by Real-Gecko** [comment]
> not reachable host caueses, i think. trying to get cpu from host info causes that.

Hmm, I cannot reproduce, can you remove problematic compute and retry.

**Comment by Real-Gecko** [comment]
> other problem in instances.html:
> `{% widthratio vm.proxy.instance.info.1 1024 1 %} MiB</td>`(Exception has occurred: VariableDoesNotExist
> Failed lookup for key [vm] in) i think this should be instance
> 
> and other in in instance_actions.html

Fixed those.

**Comment by catborise** [comment]
add a compute with fake name ip pass user. it accepts

**Comment by Real-Gecko** [comment]
I tried, I do not get the error you describe, but instead I get another one :D

**Comment by catborise** [comment]
after instance_action.html fix, click show console.: negative token generation :)
![image](https://user-images.githubusercontent.com/1725643/85290802-3ed3f000-b4a2-11ea-9abb-988f24821c7e.png)

**Comment by catborise** [comment]
forceoff power off powercycle does not work...

**Comment by catborise** [comment]
you are a good backender :)
![image](https://user-images.githubusercontent.com/1725643/85292900-434dd800-b4a5-11ea-9c57-e9789dc7d4f1.png)
i am crying :)

**Comment by Real-Gecko** [comment]
> wouldnt be nice implementation like this, if we could...

Of course it's better, but how does act during sorting and filtering? Can you try and implement it?

**Comment by Real-Gecko** [comment]
I made grouped instances index look like before 2be76c07e401d4a2875af35403ca55d550972460, but it definitely needs a rework.
![Screenshot_20200623_105735](https://user-images.githubusercontent.com/2231969/85362722-00d9d900-b541-11ea-9f52-af1a2c3145c5.png)

**Comment by catborise** [comment]
did you make any changes with console app?
i am going to work on some bugs and featues,

**Comment by Real-Gecko** [comment]
Nope.

**Comment by Real-Gecko** [comment]
> i cannot run your branch. gives error
> always i get that error


Try latest commit c2e001290587344c62025e5c08689b134797935c and see if error still exists.

**Comment by catborise** [comment]
still some errors happens not working yet because of some other errors:
![image](https://user-images.githubusercontent.com/1725643/85540522-c5620c00-b61f-11ea-8c16-1a7f25cf7bd4.png)

**Comment by Real-Gecko** [comment]
Ah yes, `storages/views.py`, will commit that.

**Comment by Real-Gecko** [comment]
b8db11f66f1dcc997c741854cdd2b9488bac9551 grab this one.
I finally started to add instances tests and with only three of them increased overall coverage from 36 to 41 percent which is goooood :D

**Comment by catborise** [comment]
yea it runs
but view not good. it should not show in the table

![image](https://user-images.githubusercontent.com/1725643/85555276-6d7ed180-b62e-11ea-8b0a-1013ccc9fe76.png)

**Comment by catborise** [comment]
> I made grouped instances index look like before [2be76c0](https://github.com/retspen/webvirtcloud/commit/2be76c07e401d4a2875af35403ca55d550972460), but it definitely needs a rework.
> ![Screenshot_20200623_105735](https://user-images.githubusercontent.com/2231969/85362722-00d9d900-b541-11ea-9f52-af1a2c3145c5.png)

we could use this datatable which has ability to group. sorting finding etc superrior than us. have you use it?
https://www.datatables.net/extensions/rowgroup/examples/initialisation/startAndEndRender.html

**Comment by Real-Gecko** [comment]
https://github.com/wenzhixin/bootstrap-table is better IMO
There's also https://django-tables2.readthedocs.io/en/latest/ which makes templates even simpler.

**Comment by Real-Gecko** [comment]
> but view not good. it should not show in the table

71d75327f21fbef126a57a378b36a5ef5584a76a
I made templates look like they did before

**Comment by catborise** [comment]
we need  "group view" or "detail view" or "tree view" like solution
https://examples.bootstrap-table.com/#extensions/treegrid.html is a suibtable one.

i do not see detail view or like that at django-tables2 .

**Comment by Real-Gecko** [comment]
> i do not see detail view or like that at django-tables2

We can override built in templates https://django-tables2.readthedocs.io/en/latest/pages/custom-rendering.html#custom-template
We can even try and mix bs-table with django-tables2, first working on client side second on server side.
I think we need to think through new layout of every view, cause now it's a bit chaotic.
Currently design like this is quite popular:
![Screenshot_20200625_122711](https://user-images.githubusercontent.com/2231969/85667284-3d433b80-b6df-11ea-983b-602281b15f83.png)
Most links are on left side menu, which is collapsible, top bar is small and does not have a lot of links. I see such layout quite often in enterprise grade products.

**Comment by catborise** [comment]
yea i like the dashboards like this. webvirtmgr has left side . webvirtcloud uses top navigation.
retspen wanted to start new project webvirtcloud v2(django rest framework + react + material ui) but for some reasons, he  choose to wait... with v2 he also switched to the dasboard style view...
nearly all dashboard templates like you specified as above...
![image](https://user-images.githubusercontent.com/1725643/85678692-6361df00-b6d1-11ea-8639-18c98ad1e570.png)
...
material-ui
![image](https://user-images.githubusercontent.com/1725643/85678894-94421400-b6d1-11ea-8ae1-089fa2deeb48.png)

i want to add a dasboard page to webvirtcloud but there was some hiccups before your reorg. may be it is more appropriate and easy adding a dashboard now before then...

**Comment by Real-Gecko** [comment]
> i want to add a dasboard page to webvirtcloud but there was some hiccups before your reorg. may be it is more appropriate and easy adding a dashboard now before then...

I'm working on instances tests, after they're done and branch is tested and merged I think we can start doing this. Instances currently has most messy layout, after splitting views I think it'll be easier to rewrite templates.

**Comment by Real-Gecko** [comment]
Was guestfsd updated for Python 3?

**Comment by catborise** [comment]
i made some changes but i am not sure if it is working or not. i never used it.

**Comment by Real-Gecko** [comment]
I'm currently trying it out and looks like it's a bit broken, will fix it.

**Comment by Real-Gecko** [comment]
ffa85d01c1f66dfc0ca26573d2f190e0411c6e16 fixed gstfsd, but it requires root privileges to run, so I'm leaving change password and add public key functionality without testing.

**Comment by catborise** [comment]
editing instance problems (retspen/master):
1. two message
![image](https://user-images.githubusercontent.com/1725643/85825093-024a1200-b78a-11ea-8154-7e65948f4ed9.png)
2. change instance console type causes error
![image](https://user-images.githubusercontent.com/1725643/85825142-2279d100-b78a-11ea-9eb9-96af974ecee4.png)

**Comment by Real-Gecko** [comment]
>     2\. change instance console type causes error

```python
console_types = appsettings.get(key="QEMU_CONSOLE_DEFAULT_TYPE").choices_as_list
```
Is the cause
must be
```python
console_types = appsettings.get(key="QEMU_CONSOLE_DEFAULT_TYPE").choices_as_list()
```

**Comment by Real-Gecko** [comment]
>     1. two message

Instances branch is almost there.

**Comment by Real-Gecko** [comment]
There is a great app https://django-guardian.readthedocs.io/en/stable/ that allows to grant users per object permissions. I've never tried it, but it looks interesting, this will grant us more granular control over instances permissions eliminating the need for UserInstance model. But this improvement is for the future, not something we must implement right now.

**Comment by catborise** [comment]
adding more features pushes to more granular permission. it will be nice more user roles then now. it should be considered after stabilizing current...

have you tried async django?
we need async capabilities for cloning instance/disk, migrate, backup... but i do not have an experience with it.
qemu provides async actions with "qemu_qpm" but not libvirt... may be django async does something?

**Comment by Real-Gecko** [comment]
> have you tried async django?

Async in Django is not fully there yet:
https://code.djangoproject.com/wiki/AsyncProject
https://speakerdeck.com/andrewgodwin/just-add-await-retrofitting-async-into-django?slide=42
So I think it's better to not bother at all for now, at least until Django 4.0.

In the mean time we'd better clean our source code a little and make it more consistent.

I finished views split for instances ae31225b0b531185f4cb39182c63a89c3fe7ef23 and looks like cleaned up all errors, anyone willing to test the branch is free to do so. In the mean time I'll continue writing auto tests for instances, we're already at 50% coverage. I'll create a PR to retspen/master as soon as no bugs remain. But there're still many places that can be improved in terms of code quality, especially in templates: it's a mess out there :D

**Comment by Real-Gecko** [comment]
https://github.com/Real-Gecko/webvirtcloud/tree/instances
link for instances testing

**Comment by Real-Gecko** [comment]
Added some more tests for instances, coverage is 55% now.

**Comment by catborise** [comment]
> https://github.com/Real-Gecko/webvirtcloud/tree/instances
> link for instances testing

i am working on reverse proxy fixing and console minor fixes. i will test it...
